### PR TITLE
test: apply strict verification flags for transaction tests and assert backwards compatibility

### PR DESCRIPTION
**Pull request by glozow**
This uses the first 4 commits of #15045, rebased and added some comments. The diff is quite large already and I want to make it easy to review, so I'm splitting it into 2 PRs (transaction and script). Script one is WIP, I'll link it when I open it.

Interpretation of scripts is dependent on the script verification flags passed in.
In tests, we should always apply **maximal** verification flags when checking that a transaction is **valid**; any additional flags should invalidate the transaction. A transaction should not be valid because we forgot to include a flag, and we should apply all flags by default.
We should apply **minimal** verification flags when asserting that a transaction is **invalid**; if verification flags are applied, removing any one of them should mean the transaction is valid.
New verify flags must be backwards compatible; tests should check backwards compatibility and apply the new flags by default. All `tx_invalid` tests should continue to be invalid with the exact same verify flags. All `tx_valid` tests that don't pass with new flags should _explicitly_ indicate that the flags need to be excluded, and fail otherwise.

1. Flip the meaning of `verifyFlags` in tx_valid.json to mean _excluded_ verification flags instead of included flags. Edit the test data accordingly.
2. Trim unneeded flags from tx_invalid.json.
3. Add check to verify that tx_valid tests have maximal flags and tx_invalid tests have minimal flags.
4. Add checks to verify that flags are soft forks (#10699) i.e. adding any flag should only decrease the number of acceptable scripts. Test by adding/removing random flags.

### CODE REVIEW

**Comment by luke-jr** [comment]
Not sure this is a good idea. Tests should ideally test one thing only, and failing due to other bugs would be annoying.

**Comment by laanwj** [comment]
Looks like one of the sanitizers finds an integer conversion/truncation problem in the changed code:
```
�[1;34;49mtest/transaction_tests.cpp(251): Leaving test case "tx_invalid"; testing time: 591590us
�[0;39;49m�[1;34;49mtest/transaction_tests.cpp(164): Entering test case "tx_valid"
test/transaction_tests.cpp:237:35: runtime error: implicit conversion from type 'unsigned long' of value 18446744073709526401 (64-bit, unsigned) to type 'unsigned int' changed the value to 4294942081 (32-bit, unsigned)
    #0 0x558b927932d2 in transaction_tests::tx_valid::test_method() /tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/transaction_tests.cpp:237:35
    #1 0x558b9278fb88 in transaction_tests::tx_valid_invoker() /tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/transaction_tests.cpp:164:1
    #2 0x558b91e8e608 in boost::detail::function::void_function_invoker0<void (*)(), void>::invoke(boost::detail::function::function_buffer&) /usr/include/boost/function/function_template.hpp:117:11
    #3 0x7f37eb6e23f1  (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x353f1)
    #4 0x7f37eb6dfc74 in boost::execution_monitor::catch_signals(boost::function<int ()> const&) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x32c74)
    #5 0x7f37eb6dfcf7 in boost::execution_monitor::execute(boost::function<int ()> const&) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x32cf7)
    #6 0x7f37eb6dfdcd in boost::execution_monitor::vexecute(boost::function<void ()> const&) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x32dcd)
    #7 0x7f37eb70d134 in boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x60134)
    #8 0x7f37eb6f05a8  (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x435a8)
    #9 0x7f37eb6f0b03  (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x43b03)
    #10 0x7f37eb6f0b03  (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x43b03)
    #11 0x7f37eb6e7939 in boost::unit_test::framework::run(unsigned long, bool) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x3a939)
    #12 0x7f37eb70bfea in boost::unit_test::unit_test_main(bool (*)(), int, char**) (/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0+0x5efea)
    #13 0x558b91dd63a6 in main /usr/include/boost/test/unit_test.hpp:63:12
    #14 0x7f37eae990b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #15 0x558b91d2bbbd in _start (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/test_bitcoin+0x21c5bbd)

SUMMARY: UndefinedBehaviorSanitizer: implicit-unsigned-integer-truncation test/transaction_tests.cpp:237:35 in 
```

**Comment by gzhao408** [comment]
@laanwj I'm looking into the issue, is there a way to run this sanitizer locally?

**Comment by gzhao408** [comment]
> Not sure this is a good idea. Tests should ideally test one thing only, and failing due to other bugs would be annoying.

@luke-jr I would think of it more as... each test sanity checks itself to make sure it's testing exactly what we want it to test. Passing/failing because the test itself is incorrect would be more annoying imo. All the current tests pass; as more tests are added, I'd think it's a good idea to have this check.

**Comment by gzhao408** [comment]
A little bit quiet here... @laanwj and @practicalswift you both left reviews on the original PR, if you have time I'd appreciate a look here as well :)

**Comment by laanwj** [comment]
Code review ACK, thanks for solving the sanitizer bug.

I would like @sipa @MarcoFalke or someone else close with the verification testing code to look at this and give concept ACK.

**Comment by benthecarman** [comment]
Concept ACK

File: src/test/transaction_tests.cpp [hunk 1]
```diff
@@ -240,27 +306,34 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
...
-                    amount = mapprevOutValues[tx.vin[i].prevout];
-                }
-                const CScriptWitness *witness = &tx.vin[i].scriptWitness;
-                fValid = VerifyScript(tx.vin[i].scriptSig, mapprevOutScriptPubKeys[tx.vin[i].prevout],
-                                      witness, verify_flags, TransactionSignatureChecker(&tx, i, amount, txdata), &err);
+            // Not using FillFlags() in the main test, in order to detect invalid verifyFlags combination
+            if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, verify_flags[0], txdata, strTest, /* expect_valid */ false))
+                BOOST_ERROR("Tx unexpectedly passed: " << strTest);
+            // Check that flags are minimal: transaction should succeed if any set flags are unset.
+            for (size_t i = 0; i < mapFlagNames.size(); i++)
```
Comment (benthecarman) [inline src/test/transaction_tests.cpp:321]: nit: ++i

File: src/test/transaction_tests.cpp [hunk 2]
```diff
@@ -240,27 +306,34 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
...
+                // Backwards compatibility: Adding a flag should not validate an invalid transaction
+                unsigned int flags = FillFlags(verify_flags[0] | (1U << i));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ false))
+                    BOOST_ERROR("Tx unexpectedly passed with flag " << ToString(i) << " set: " << strTest);
+                // Backwards compatibility: Adding some random flags should not validate an invalid transaction
+                flags = FillFlags(verify_flags[0] | (unsigned int)InsecureRandBits(mapFlagNames.size()));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ false))
+                    BOOST_ERROR("Tx unexpectedly passed with random flags " << ToString(flags) << ": " << strTest);
+            }
+            for (size_t f = 1; f < verify_flags.size(); f++) {
```
Comment (benthecarman) [inline src/test/transaction_tests.cpp:332]: nit: ++f

File: src/test/transaction_tests.cpp [hunk 3]
```diff
@@ -104,10 +169,9 @@ BOOST_AUTO_TEST_CASE(tx_valid)
     // or [[[prevout hash, prevout index, prevout scriptPubKey], [input 2], ...],"], serializedTransaction, verifyFlags
     // ... where all scripts are stringified scripts.
     //
-    // verifyFlags is a comma separated list of script verification flags to apply, or "NONE"
+    // verifyFlags is a comma separated list of MINIMAL script verification flags to exclude, or "NONE"
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:172]: I don't think the format of tx_valid.json should be documented here _and_ in the file itself. It'd be better to consolidate this documentation to just the json file.

No need to capitalize MINIMAL.

File: src/test/transaction_tests.cpp [hunk 4]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
...
+        flags &= ~(unsigned int)SCRIPT_VERIFY_CLEANSTACK;
+    }
+    else if (!(flags & SCRIPT_VERIFY_P2SH)) {
+        flags &= ~(unsigned int)SCRIPT_VERIFY_WITNESS;
+        flags &= ~(unsigned int)SCRIPT_VERIFY_CLEANSTACK;
+    }
+    return flags;
+}
+
+unsigned int FillFlags(unsigned int flags)
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:150]: Would a better interface be to take a reference and update it in place?

File: src/test/transaction_tests.cpp [hunk 5]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
...
+{
+    if (flags & SCRIPT_VERIFY_WITNESS) {
+        flags |= SCRIPT_VERIFY_P2SH;
+    }
+    else if (flags & SCRIPT_VERIFY_CLEANSTACK) {
+        flags |= SCRIPT_VERIFY_P2SH;
+        flags |= SCRIPT_VERIFY_WITNESS;
+    }
+    return flags;
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:159]: Perhaps clearer:

```suggestion
    // CLEANSTACK implies WITNESS
    if (flags & SCRIPT_VERIFY_CLEANSTACK) flags |= SCRIPT_VERIFY_WITNESS;
    
    // WITNESS implies P2SH (and transitively CLEANSTACK implies P2SH)
    if (flags & SCRIPT_VERIFY_WITNESS) flags |= SCRIPT_VERIFY_P2SH;
```

File: src/test/transaction_tests.cpp [hunk 6]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
     return ret.substr(0, ret.size() - 1);
 }
 
+bool CheckTxScripts(const CTransaction& tx, const std::map<COutPoint, CScript>& mapprevOutScriptPubKeys, const std::map<COutPoint, int64_t>& mapprevOutValues, const unsigned int& flags, const PrecomputedTransactionData& txdata, const std::string& strTest, const bool expect_valid)
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:103]: Please split this line (and perhaps comment what this function is doing)

File: src/test/transaction_tests.cpp [hunk 7]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
...
+{
+    bool tx_valid = true;
+    ScriptError err = expect_valid ? SCRIPT_ERR_UNKNOWN_ERROR : SCRIPT_ERR_OK;
+    for (unsigned int i = 0; i < tx.vin.size() && tx_valid; i++)
+    {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:108]: join

File: src/test/transaction_tests.cpp [hunk 8]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
...
+        const CAmount amount = mapprevOutValues.count(tx.vin[i].prevout) ? mapprevOutValues.at(tx.vin[i].prevout) : 0;
+        try
+        {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:111]: join

File: src/test/transaction_tests.cpp [hunk 9]
```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
...
+            tx_valid = VerifyScript(tx.vin[i].scriptSig, mapprevOutScriptPubKeys.at(tx.vin[i].prevout), &tx.vin[i].scriptWitness, flags, TransactionSignatureChecker(&tx, i, amount, txdata), &err);
+        }
+        catch (...)
+        {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:115]: join

File: src/test/transaction_tests.cpp [hunk 10]
```diff
@@ -157,24 +221,28 @@ BOOST_AUTO_TEST_CASE(tx_valid)
...
-                }
-                unsigned int verify_flags = ParseScriptFlags(test[2].get_str());
-                const CScriptWitness *witness = &tx.vin[i].scriptWitness;
-                BOOST_CHECK_MESSAGE(VerifyScript(tx.vin[i].scriptSig, mapprevOutScriptPubKeys[tx.vin[i].prevout],
-                                                 witness, verify_flags, TransactionSignatureChecker(&tx, i, amount, txdata), &err),
-                                    strTest);
-                BOOST_CHECK_MESSAGE(err == SCRIPT_ERR_OK, ScriptErrorString(err));
+            // Do not use TrimFlags() in the main test in order to detect invalid verifyFlags combination in the test file.
+            if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~verify_flags[0], txdata, strTest, /* expect_valid */ true))
+                BOOST_ERROR("Tx unexpectedly failed: " << strTest);
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:228]: braces

File: src/test/transaction_tests.cpp [hunk 11]
```diff
@@ -157,24 +221,28 @@ BOOST_AUTO_TEST_CASE(tx_valid)
...
+            // Check that flags are maximal: transaction should fail if any unset flags are set.
+            for (size_t i = 0; i < mapFlagNames.size(); i++)
+            {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:231]: join

File: src/test/transaction_tests.cpp (hunk 7 above)
Comment (jnewbery) [inline src/test/transaction_tests.cpp:108]: consider making an alias reference for `tx.vin[i]` here, since you'll use it a lot below.

File: src/test/transaction_tests.cpp [hunk 12]
```diff
@@ -59,10 +59,12 @@ static std::map<std::string, unsigned int> mapFlagNames = {
     {std::string("CONST_SCRIPTCODE"), (unsigned int)SCRIPT_VERIFY_CONST_SCRIPTCODE},
 };
 
-unsigned int ParseScriptFlags(std::string strFlags)
+std::vector<unsigned int> ParseScriptFlags(std::string strFlags)
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:62]: This is a really weird function. It'll now return a vector of:
- first item is the sum of all the flags
- then a list of the sum of all except one flag for each flag.

I think the function should be left unchanged and the client do whatever it needs with the sum of all the flags. At the very least, the function should be documented.

File: src/test/transaction_tests.cpp [hunk 13]
```diff
@@ -157,24 +221,28 @@ BOOST_AUTO_TEST_CASE(tx_valid)
...
+            // Do not use TrimFlags() in the main test in order to detect invalid verifyFlags combination in the test file.
+            if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~verify_flags[0], txdata, strTest, /* expect_valid */ true))
+                BOOST_ERROR("Tx unexpectedly failed: " << strTest);
+            // Check that flags are maximal: transaction should fail if any unset flags are set.
+            for (size_t i = 0; i < mapFlagNames.size(); i++)
+            {
+                // Backwards compatibility: Removing a flag should not invalidate a valid transaction
+                unsigned int flags = TrimFlags(~(verify_flags[0] | (1U << i)));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true))
+                    BOOST_ERROR("Tx unexpectedly failed with flag " << ToString(i) << " unset: " << strTest);
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:235]: braces

File: src/test/transaction_tests.cpp [hunk 14]
```diff
@@ -157,24 +221,28 @@ BOOST_AUTO_TEST_CASE(tx_valid)
...
+                // Backwards compatibility: Removing some random flags should not invalidate a valid transaction
+                flags = TrimFlags(~(verify_flags[0] | (unsigned int)InsecureRandBits(mapFlagNames.size())));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true))
+                    BOOST_ERROR("Tx unexpectedly failed with random flags " << ToString(flags) << ": " << strTest);
+            }
+            for (size_t f = 1; f < verify_flags.size(); f++)
+            {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:242]: join

File: src/test/transaction_tests.cpp [hunk 15]
```diff
@@ -157,24 +221,28 @@ BOOST_AUTO_TEST_CASE(tx_valid)
...
+                // Backwards compatibility: Removing a flag should not invalidate a valid transaction
+                unsigned int flags = TrimFlags(~(verify_flags[0] | (1U << i)));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true))
+                    BOOST_ERROR("Tx unexpectedly failed with flag " << ToString(i) << " unset: " << strTest);
+                // Backwards compatibility: Removing some random flags should not invalidate a valid transaction
+                flags = TrimFlags(~(verify_flags[0] | (unsigned int)InsecureRandBits(mapFlagNames.size())));
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true))
+                    BOOST_ERROR("Tx unexpectedly failed with random flags " << ToString(flags) << ": " << strTest);
+            }
+            for (size_t f = 1; f < verify_flags.size(); f++)
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:241]: Use range-based loop?

File: src/test/transaction_tests.cpp [hunk 16]
```diff
@@ -240,27 +306,34 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
 
             TxValidationState state;
             fValid = CheckTransaction(tx, state) && state.IsValid();
+            if (!fValid) {
```
Comment (jnewbery) [inline src/test/transaction_tests.cpp:309]: join these lines and don't use the local variable `fValid` (it means 'the test file was malformed' above and 'the transaction failed validation' here).

File: src/test/transaction_tests.cpp (hunk 2 above)
Comment (jnewbery) [reply src/test/transaction_tests.cpp:332]: or just use range-based for loop.

**Review by jnewbery — COMMENTED** [review]
I've left lots of style suggestions, which should be pretty easy to resolve.

It's very difficult to review this PR, since it's mixing refactors and fixes into the same commits. For example, the first commit is changing the format of the tx_valid.json file _and_ making changes/fixes to the tests. I think it'd be far easier to review those different changes if they were split out into individual commits. The same is true for the other commits in this branch - each one is doing too much, making review more difficult than it needs to ne.

**Review by murchandamus — COMMENTED** [review]
Concept ACK

TBH, I could use a bit more explanation in the fourth commit message "Verify that all validation flags are backward compatible". Could you go a bit more into detail what this is doing and why?

File: src/test/transaction_tests.cpp (hunk 15 above)
Comment (jonatack) [reply src/test/transaction_tests.cpp:241]: And otherwise, prefer ++f to f++

File: src/test/transaction_tests.cpp (hunk 14 above)
Comment (jonatack) [reply src/test/transaction_tests.cpp:242]: @glozow if you run clang-format on your changes, it can avoid the need for all the formatting feedback

File: src/test/transaction_tests.cpp (hunk 6 above)
Comment (jonatack) [reply src/test/transaction_tests.cpp:103]: @glozow I find https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rf-in useful for when to pass "in" parameters by value versus by reference to const

Comment (jonatack) [reply src/test/transaction_tests.cpp:103]: > Have line length guidelines ever been considered for this codebase? innocent

For the C++ part, the longer the better? (After many years of 80 or 100 char line limits, this surprised me at first, but for the C++ code I now actually prefer the lines to be unbroken more often than not.)

For the Python code, ISTM the unofficial practice is not PEP8 (79 chars) but more like 120 or 160 chars.

File: src/test/transaction_tests.cpp (hunk 14 above)
Comment (glozow) [reply src/test/transaction_tests.cpp:242]: oops 😅 thanks for advice @jonatack

Comment (michaelfolkson) [reply src/test/transaction_tests.cpp:242]: https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#clang-format-diffpy

File: src/test/transaction_tests.cpp (hunk 4 above)
Comment (glozow) [reply src/test/transaction_tests.cpp:150]: maybe not, since we're using it like
`CheckTxScripts(... FillFlags(flags) ...)`
So doing it in place would mean we need an extra line there

File: src/test/transaction_tests.cpp (hunk 3 above)
Comment (glozow) [reply src/test/transaction_tests.cpp]: consolidated in the json since that's where any new tests would be added

File: src/test/transaction_tests.cpp (hunk 5 above)
Comment (glozow) [reply src/test/transaction_tests.cpp]: indeed, cleaner :)

File: src/test/data/tx_valid.json [hunk 17]
```diff
@@ -2,421 +2,421 @@
...
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0x05 0x0000000000 CHECKSEQUENCEVERIFY 1"]],
-"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
+"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000", "MINIMALDATA,CLEANSTACK"],
 
 ["The argument can be calculated rather than created directly by a PUSHDATA"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4194303 1ADD CHECKSEQUENCEVERIFY 1"]],
-"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000040000100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4194304 1SUB CHECKSEQUENCEVERIFY 1"]],
-"020000000100010000000000000000000000000000000000000000000000000000000000000000000000ffff00000100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
+[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4194303 1ADD CHECKSEQUENCEVERIFY"]],
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/test/data/tx_valid.json:307]: pulled them into their own commits

File: src/test/data/tx_invalid.json [hunk 18]
```diff
@@ -273,7 +273,7 @@
...
 ["0000000000000000000000000000000000000000000000000000000000000100", 2, "0x51", 3000]],
-"0100000000010300010000000000000000000000000000000000000000000000000000000000000000000000ffffffff000100000000000000000000000000000000000000000000000000000000000001000000000100000000010000000000000000000000000000000000000000000000000000000000000200000000ffffffff00000248304502210091b32274295c2a3fa02f5bce92fb2789e3fc6ea947fbe1a76e52ea3f4ef2381a022079ad72aefa3837a2e0c033a8652a59731da05fa4a813f4fc48e87c075037256b822103596d3451025c19dbbdeb932d6bf8bfb4ad499b95b6f88db8899efac102e5fc710000000000", "P2SH,WITNESS"],
+"0100000000010300010000000000000000000000000000000000000000000000000000000000000000000000ffffffff000100000000000000000000000000000000000000000000000000000000000001000000000100000000010000000000000000000000000000000000000000000000000000000000000200000000ffffffff03e8030000000000000151d0070000000000000151b80b0000000000000151000248304502210091b32274295c2a3fa02f5bce92fb2789e3fc6ea947fbe1a76e52ea3f4ef2381a022079ad72aefa3837a2e0c033a8652a59731da05fa4a813f4fc48e87c075037256b822103596d3451025c19dbbdeb932d6bf8bfb4ad499b95b6f88db8899efac102e5fc710000000000", "P2SH,WITNESS"],
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/test/data/tx_invalid.json:277]: also 👍

File: src/test/transaction_tests.cpp (hunk 6 above)
Comment (glozow) [reply src/test/transaction_tests.cpp]: I eyeballed ~100 chars, hopefully that's ok

**Comment by glozow** [comment]
Thanks for the review @laanwj @jnewbery @benthecarman @Xekyo @jonatack :) addressed your comments and split the PR up into more, dedicated commits. CI is green 🟢 ready for review again!

File: src/test/transaction_tests.cpp [hunk 19]
```diff
@@ -95,14 +92,83 @@ std::string FormatScriptFlags(unsigned int flags)
     return ret.substr(0, ret.size() - 1);
 }
 
+/*
+* Check that the input scripts of a transaction are valid/invalid as expected.
+*/
+bool CheckTxScripts(const CTransaction& tx, const std::map<COutPoint, CScript>& mapprevOutScriptPubKeys,
+    const std::map<COutPoint, int64_t>& mapprevOutValues, unsigned int flags,
```
Comment (Xekyo) [inline src/test/transaction_tests.cpp:100]: Nit: map_p_rev. Also, I seem to remember that new variables should preferably use snake case.

**Review by murchandamus — COMMENTED** [review]
Looks good to me, just the one nit from me at the moment.

File: src/test/transaction_tests.cpp (hunk 19 above)
Comment (jonatack) [reply src/test/transaction_tests.cpp:100]: That's true (doc/developer-notes.md)

File: src/test/transaction_tests.cpp [hunk 20]
```diff
@@ -95,20 +93,83 @@ std::string FormatScriptFlags(unsigned int flags)
...
+    const PrecomputedTransactionData& txdata, const std::string& strTest, bool expect_valid)
+{
+    bool tx_valid = true;
+    ScriptError err = expect_valid ? SCRIPT_ERR_UNKNOWN_ERROR : SCRIPT_ERR_OK;
+    for (unsigned int i = 0; i < tx.vin.size() && tx_valid; i++) {
```
Comment (jonatack) [inline src/test/transaction_tests.cpp:105]: drive-by nit if you retouch: s/i++/++i/ (see developer-notes.md)

Comment (glozow) [reply src/test/transaction_tests.cpp:105]: GAH missed that 😭

File: src/test/transaction_tests.cpp (hunk 19 above)
Comment (jonatack) [reply src/test/transaction_tests.cpp:100]: `flags` and `expect_valid` now passed by value, thanks :+1:

Comment (glozow) [reply src/test/transaction_tests.cpp:100]: o tru

**Comment by glozow** [comment]
Attempting to revive this again 😅 🙏 
Rebased on master since it's been a while and applied the style suggestions from @jonatack and @Xekyo. I had also forgotten to add the original author to the intermediate commits when I split up the changes - my sincerest apologies - that's fixed now.

**Comment by laanwj** [comment]
To test the tests I made a small change to `tx_valid.json` (basically reverting the initial commits):
```patch
 ["An ADD producing a 5-byte result that sets CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 65536 ADD CHECKSEQUENCEVERIFY"]],
+[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 65536 CHECKSEQUENCEVERIFY"]],
```
The good thing is that it easily detects this. However, an avalanche of 63 failures for a single error might be a bit overkill :sweat_smile: It does report the JSON of the failed test, which is good!

The following change, however
```patch
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 65536 ADD CHECKSEQUENCEVERIFY"]],
-"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000", "NONE"],
+"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
```
raises an assertion
```
test_bitcoin: …/bitcoin/src/script/interpreter.cpp:2044: bool VerifyScript(const CScript &, const CScript &, const CScriptWitness *, unsigned int, const BaseSignatureChecker &, ScriptError *): Assertion `(flags & SCRIPT_VERIFY_P2SH) != 0' failed.
```
Then reports 7 failures in different places in the code. It does not report which specific test in the JSON failed.

Not sure if these are a problem, I mean, the tests fail when they're supposed to fail, but just thought I'd report it.

**Comment by glozow** [comment]
@laanwj Good point, it's not very helpful if `VerifyFlags` just throws and we don't know which test failed. I've added a check using `FillFlags` to throw a "Bad test flags" when there's an invalid combination of flags given in tx_valid.json (https://github.com/bitcoin/bitcoin/pull/19698/commits/a260c22cad0672dda11f42f649ebdc7cfa53b16a).

I tested by adding something very similar to what you did in tx_valid.json (these would result in a `CLEANSTACK` that's missing `P2SH` and `WITNESS`, respectively, which we know to be invalid/non-backwards-compatible combinations:
```
["BAD TEST - excluding P2SH or WITNESS but not CLEANSTACK creates an invalid combination of flags"],
[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 65536 ADD CHECKSEQUENCEVERIFY"]],
"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000", "P2SH"],
[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 4259840 ADD CHECKSEQUENCEVERIFY"]],
"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000040000100000000000000000000000000", "WITNESS"],
```

And it should print which test failed:
```
test/transaction_tests.cpp:228: error: in "transaction_tests/tx_valid": Bad test flags: [[["0000000000000000000000000000000000000000000000000000000000000100",0,"2147483647 65536 ADD CHECKSEQUENCEVERIFY"]],"020000000100010000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000","P2SH"]
Assertion failed: ((flags & SCRIPT_VERIFY_P2SH) != 0), function VerifyScript, file script/interpreter.cpp, line 2044.
```
Let me know what you think?

**Comment by achow101** [comment]
ACK 5786a818e1a96bc1dd65b0e81b05998876357a74

**Comment by laanwj** [comment]
> Let me know what you think?

Looks good to me now!

ACK 5786a818e1a96bc1dd65b0e81b05998876357a74

File: src/test/transaction_tests.cpp [hunk 21]
```diff
@@ -96,14 +93,84 @@ std::string FormatScriptFlags(unsigned int flags)
...
+    return flags;
+}
+
+// Return valid flags that are all except one flag for each flag
+std::vector<unsigned int> ExcludeIndividualFlags(unsigned int flags)
+{
+    std::vector<unsigned int> flags_combos;
+    for (unsigned int i = 0; i < mapFlagNames.size(); ++i) {
+        const unsigned int flags_excluding_i = TrimFlags(flags & ~(1U << i));
+        if (flags != flags_excluding_i && std::find(flags_combos.begin(), flags_combos.end(), flags_excluding_i) != flags_combos.end()) {
```
Comment (MarcoFalke) [inline src/test/transaction_tests.cpp:161]: I don't understand what this is supposed to do. The method just returns an empty vector.

**Review by maflcko — COMMENTED** [review]
Concept ACK, but I don't think this works

File: src/test/transaction_tests.cpp (hunk 21 above)
Comment (glozow) [reply src/test/transaction_tests.cpp:161]: I don't think it always returns an empty vector? I'm playing around with it, should only return empty if `flags=0`.

What it's supposed to do: given a set of verify flags `flags`, and all the flags in existence from `mapFlagNames` and exclude each one from `flags` (granted it's a valid combination and doesn't just result in the same flags).

So for example if it's given 10001100 and there are 8 total flags, it'll return [00001100, 10000100, 10001000] if they're all valid combinations. If the input is 0, it should return an empty vector.

Comment (glozow) [reply src/test/transaction_tests.cpp:161]:
```suggestion
        if (flags != flags_excluding_i && std::find(flags_combos.begin(), flags_combos.end(), flags_excluding_i) == flags_combos.end()) {
```
Yeah, it should be this. Should not be found in `flags_combos`. I'll open a PR to address this

File: src/test/data/tx_invalid.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (glozow) [reply src/test/data/tx_invalid.json:277]: I had interpreted that I was supposed to put the 2 tx_invalid test fixes into a separate commit (they were initially grouped with the "apply minimal flags" commit) so I thought I had done this, oops! Do you have the test failing? Should I do something to fix this now?

> Also, I am wondering why exactly 3 outputs? Wouldn't one suffice?

Yeah I think so. I just pulled what was on jl2012's branch 😅

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/data/tx_invalid.json:277]: It is not possible to rewrite our git history, so the failing commit will need to be kept as is.
